presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +494,53 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        // 1. normal aggregation
+        if (!preGroupedHashStrategy.isPresent()) {
+            unfinishedWork = aggregationBuilder.processPage(page);
+            return;
+        }
+
+        // 2. segmented aggregation
+        if (firstUnfinishedSegment == null) {
```

Comment:
Same here. Whether to use pre-grouped channel or not is actually determined at the beginning of the execution or planning phase. There is no need to check if/else for every page input. Instead, it would be good to abstract the design a bit. For example, can we have a base abstract hash operator with two implementations: Hash and Segmented. Then we can leave the branches within the different implementations

Reply:
> Whether to use pre-grouped channel or not is actually determined at the beginning of the execution or planning phase. There is no need to check if/else for every page input.

Great point.
> can we have a base abstract hash operator with two implementations: Hash and Segmented.

I thought about this idea as well before implementing the PR. Not sure if it's worth it to restructure the whole operator given that Segmented aggregation is essentially still Hash aggregation just with a few tricks. The difference is not that big. WDYT? @kewang1024